plutus-core/cost-model/CostModelGeneration.md

```diff
@@ -59,48 +62,380 @@ Generating a cost model for CPU time involves a number of steps.
+times incur large charges which penalise excessive computation.  Some
+experimentation may be required to achieve this, and it may not always be
+possible to satisfy both goals simultaneously.  In such cases it may be
+necessary to sacrifice some accuracy in order to guarantee security.
+
+
+#### Step 6: add code to read the costing function from R into Haskell 
+
+Next we have to update the code which converts benchmarking results into JSON
+models.  Go to
+[`CreateBuiltinCostModel`](./create-cost-model/CreateBuiltinCostModel.hs) and add
+entries for the new builtin in builtinCostModelNames
+
+```
+  , paramXorByteString                   = "xorByteStringModel"
+```
+(Getting the string wrong here, for example putting "xorByteString" instead will
+give `parse error (not enough input) at ""`. Errors will occur whenever the
+Haskell code attempts to read something from an R object that doesn't actually
+occur in the object, and they can sometimes be quite cryptic.)
+
+Also add a new clause in [`CreateBuiltinCostModel`](./create-cost-model/CreateBuiltinCostModel.hs):
+
+```
+    paramXorByteString                   <- getParams xorByteString paramXorByteString
+```
+
+and a function to extract the cost parameters for the R code.  This should be modelled on the existing
+functions at the end of the file:
+
+```
+xorByteString :: MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+xorByteString cpuModelR = do
+  cpuModel <- ModelTwoArgumentsMinSize <$> readModelMinSize cpuModelR
+  let memModel = ModelTwoArgumentsMinSize $ ModelMinSize 0 1
+  pure $ CostingFun (cpuModel) memModel
+```
+
+The CPU costing function is obtained by running the R code, but the memory usage
+costing function is defined statically here.  Memory usage costing functions
+only account for memory retained after the function has returned and not for any
+working memory that may be allocated during its execution.  Typically this means
+that the memory costing function should measure the size of the object returned
+by the builtin.  For our `xorByteString` implementation, if the arguments have
+sizes `m` and `n` then the result will have size `min(m,n)` so we define the memory
+costing function to be `(m,n) -> 0 + 1*min(m,n)`.
+
+
+#### Step 7: update the R code
+
+We now have to extend the R code in [`models.R`](./data/models.R).  Firstly, add
+an entry for the arity of the builtin in the `arity` function:
+
+```
+   arity <- function(name) {
+       switch (name,
+           "AddInteger" = 2,
+           ...
+           "XorByteString" = 2,
+           ...
+           )
+```
+
+Now add a function to infer coefficients for the CPU costing function from
+benchmarking data.  In the case of `xorByteString` we assume that the time taken
+will be linear in the minimum of the sizes of the arguments (ie, the arguments
+of the new builtin).  It is often worthwhile to plot the benchmark data and
```

Comment:
This seems quite important, is it worth elaborating further? I'm wondering:
- Instead of manually plotting the benchmark data and inspecting it, is it possible to automatically evaluate the assumed model using some metrics, like mean squared error or something?
- How exactly does one plot the benchmark data, do we have a specific tool for that?

Reply:
> Instead of manually plotting the benchmark data and inspecting it, is it possible to automatically evaluate the assumed model using some metrics, like mean squared error or something?

That's what the R code does: you give it a basic shape ("linear in the sum of the sizes of the inputs" or something) and then it does some regression on the benchmark results to work out what the actual numbers should be.  It's useful (at least at first) to experiment with the data to make sure that it looks like you're expecting it to though, rather than just letting R fit a model to the data.  If you get the model wrong it'll happy fit a straight line to some quadratic data, for example.

> How exactly does one plot the benchmark data, do we have a specific tool for that?

I do that in R, but I suppose you could use whatever tool you happen to be familiar with.

Reply:
> you give it a basic shape ("linear in the sum of the sizes of the inputs" or something)

Can you make it automatically choose the best shape among all options based on some metrics? One of them should have a much better fit than others.

Reply:
> Can you make it automatically choose the best shape among all options based on some metrics? One of them should have a much better fit than others.

I think that's probably quite a difficult problem in general.  We do have a limited number of basic shapes so I suppose we could try them all. That still seems a little dangerous though: maybe your data doesn't really fit any of them and you need to introduce a new one, but how would you know that? Looking at a picture seems like the easiest way.  R does calculate a number of goodness-of-fit measures when you fit a model to your data, but they can be quite difficult to interpret, especially if you've got multiple arguments.  I'll think about that some more.